<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>3&nbsp; Turing patterns, repetitive patterning – Modeling Life</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./pattern_practical_3.html" rel="next">
<link href="./pattern_practical_1.html" rel="prev">
<link href="./images/favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./pattern_intro_text.html">I) Pattern formation</a></li><li class="breadcrumb-item"><a href="./pattern_practical_2.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Turing patterns, repetitive patterning</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./images/android-chrome-512x512.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Modeling life</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false">
 <span class="menu-text">Course information</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Modeling life</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./general.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">General course info</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slides</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro_to_python.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">0) Introduction to Python</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">I) Pattern formation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Pattern formation (intro)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_practical_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Gradients and segments</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_practical_2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Turing patterns, repetitive patterning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_practical_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Clock-and-wavefront patterning</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false">
 <span class="menu-text">II) Morphogenesis</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./morpho_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Morphogenesis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./morpho_cpm_practical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell sorting by differential adhesion</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">III) Cell differentiation</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differentiation_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell differentiation introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differentiation_practical_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Gene regulation in time</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differentiation_practical_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Gene regulation in space</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false">
 <span class="menu-text">IV) Environment</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evdev_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Environmental inputs</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evdev_practical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Modeling Thermomorphogenesis in Plants</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false">
 <span class="menu-text">V) Evolution</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to evolution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_practical_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Sticking together</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_practical_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">What gets selected?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_practical_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Public goods</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="-1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul class="collapse">
  <li><a href="#section" id="toc-section" class="nav-link active" data-scroll-target="#section"></a></li>
  <li><a href="#equivalence-of-turing-models" id="toc-equivalence-of-turing-models" class="nav-link" data-scroll-target="#equivalence-of-turing-models"><span class="header-section-number">3.1</span> Equivalence of Turing models</a></li>
  <li><a href="#patterning-in-a-fixed-domain-size" id="toc-patterning-in-a-fixed-domain-size" class="nav-link" data-scroll-target="#patterning-in-a-fixed-domain-size"><span class="header-section-number">3.2</span> Patterning in a fixed domain size</a></li>
  <li><a href="#patterning-in-a-growing-domain-size" id="toc-patterning-in-a-growing-domain-size" class="nav-link" data-scroll-target="#patterning-in-a-growing-domain-size"><span class="header-section-number">3.3</span> Patterning in a growing domain size</a></li>
  <li><a href="#patterning-along-two-growth-axes" id="toc-patterning-along-two-growth-axes" class="nav-link" data-scroll-target="#patterning-along-two-growth-axes"><span class="header-section-number">3.4</span> Patterning along two growth axes</a></li>
  <li><a href="#making-a-virtual-tissue-grow" id="toc-making-a-virtual-tissue-grow" class="nav-link" data-scroll-target="#making-a-virtual-tissue-grow"><span class="header-section-number">3.5</span> Making a virtual tissue grow</a></li>
  <li><a href="#spatial-modulation-of-parameters-k4-and-k7" id="toc-spatial-modulation-of-parameters-k4-and-k7" class="nav-link" data-scroll-target="#spatial-modulation-of-parameters-k4-and-k7"><span class="header-section-number">3.6</span> Spatial modulation of parameters k4 and k7</a></li>
  <li><a href="#hoxd13-and-fgf" id="toc-hoxd13-and-fgf" class="nav-link" data-scroll-target="#hoxd13-and-fgf"><span class="header-section-number">3.7</span> Hoxd13 and FGF</a></li>
  <li><a href="#shape-makes-the-tissue" id="toc-shape-makes-the-tissue" class="nav-link" data-scroll-target="#shape-makes-the-tissue"><span class="header-section-number">3.8</span> Shape makes the tissue</a></li>
  <li><a href="#master-students-exercise" id="toc-master-students-exercise" class="nav-link" data-scroll-target="#master-students-exercise"><span class="header-section-number">3.9</span> Master students exercise</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./pattern_intro_text.html">I) Pattern formation</a></li><li class="breadcrumb-item"><a href="./pattern_practical_2.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Turing patterns, repetitive patterning</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Turing patterns, repetitive patterning</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="section" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="section"></h2>
<p>A foundational work in Theoretical Biology is Alan Turing’s paper <em>“The Chemical Basis of Morphogenesis”</em> (<span class="citation" data-cites="turing1990chemical">Turing (<a href="references.html#ref-turing1990chemical" role="doc-biblioref">1990</a>)</span>). In this work, Turing mathematically demonstrates a plausible mechanism whereby a reaction-diffusion system can lead to the formation of a spatially-patterned solution in steady state. Mathematically and conceptually, Turing’s key innovation was considering diffusion-driven instability. That is: A non-patterned steady state solution becomes unstable in the presence of diffusion, leading to the formation of a spatially patterned state.</p>
<p>Turing-type patterns emerge when a system exhibits short-ranged activation and long-ranged inhibition. The most well-known realization of a Turing reaction-diffusion system is the combination of a slow-diffusing activator and fast-diffusing inhibitor. Many more examples exist which can be reduced to a mathematically equivalent form. In this practical, we will explore a 3-component Turing system developed by Raspopovic and coauthors to explain the patterning of fingers on the paws of mice (<span class="citation" data-cites="raspopovic2014digit">Raspopovic et al. (<a href="references.html#ref-raspopovic2014digit" role="doc-biblioref">2014</a>)</span>).</p>
<div class="columns">
<div class="column" style="width:40%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="images/0-hamster.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Credit: the internet."><img src="images/0-hamster.png" class="img-fluid figure-img" style="width:40.0%" alt="Credit: the internet."></a></p>
<figcaption>Credit: the internet.</figcaption>
</figure>
</div>
</div><div class="column" style="width:60%;">
<p>In this practical you will work with a total of 6 different python codes. Please note that each code simply differs by an extension or modification from the previous code, so that in the end you will have a long script that performs all the steps in one go.</p>
</div>
</div>
<hr>
</section>
<section id="equivalence-of-turing-models" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="equivalence-of-turing-models"><span class="header-section-number">3.1</span> Equivalence of Turing models</h2>
<p>The model proposed by Raspopovic et al.&nbsp;for generating Turing patterns underlying limb bone modeling consists of 3 instead of 2 interacting variables. We will refer to this model as “BSW” model, as it represents interactions between Bmp, Sox9, and Wnt.</p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.1 (Conceptual thinking)</strong></span> Study the schemes below. The arrows represent activating or inhibiting interactions, while the wavy lines represent diffusion of components. Simplify this 3-variable model scheme into a 2 variable model scheme (just by drawing the interactions, no need to write equations) and explain how it satisfies the Turing conditions discussed in the lecture. Remember that for Turing patterns we need an activator an inhibitor.</p>
<p>In a 2 variable system the activator needs to be a single autoactivating molecule. When we have more variables, the positive feedback can be achieved differently (indirectly). So think which two variables would be logical to combine into one to get a 2 variable model..</p>
<div class="columns">
<div class="column" style="width:40%;">
<div id="fig-turing" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-turing-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="images/two_component_turing.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;3.1: Schemes and numerical solutions of two prototypical 2-component Turing models"><img src="images/two_component_turing.png" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-turing-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.1: Schemes and numerical solutions of two prototypical 2-component Turing models
</figcaption>
</figure>
</div>
</div><div class="column" style="width:60%;">
<div id="fig-wntbmpsox" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-wntbmpsox-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="images/wntbmpsox.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;3.2: Scheme of the BSW model"><img src="images/wntbmpsox.png" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-wntbmpsox-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.2: Scheme of the BSW model
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> “fuse” Sox9 and Wnt nodes to get Substrate-Depletion model <br>a) Sox9 inhibits Wnt which inhibits Sox9 -&gt; summarize as self-activation <br>b) Wnt self-inhibition arrow -&gt; part of self-activation in reduced scheme <br>c) Wnt slow diffusion -&gt; Sox9 slow diffusion<br><br> Turing instability requires short-range activation and long-range inhibition. <br><br> In the substrate-depletion model, the growth of the slow-diffusing activator A is constrained by the local loss of fast-diffusing substrate S. Around an A-enriched domain a halo of S will form, which will induce nearby A domains to grow, generating the alternating spatial pattern.<br><br> In the BSW model, Sox9 indirectly locally self-activates by inhibiting its slow-diffusing inhibitor Wnt, but the growth of Sox9 domains is constrained by depletion of the fast-diffusing Bmp (playing the role of substrate in this model). Around Sox9-enriched domains a halo of Bmp will form and induce the generation of more Sox9 domains in the surroundings.</p>
</blockquote>
</section>
<section id="patterning-in-a-fixed-domain-size" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="patterning-in-a-fixed-domain-size"><span class="header-section-number">3.2</span> Patterning in a fixed domain size</h2>
<p>The code in the script <code>01__digits_squarehomogeneoustissue.py</code> simulates the model in a 2D square tissue of constant size. This setup can be considered as a small region of a large Petri dish in which mesenchymal cells have been plated with the correct chemicals to undergo bone formation (Supplementary Figure S3 in the paper).</p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.2 (Biology)</strong></span> Play with parameters <span class="math inline">\(k4\)</span> (Sox9 effect on BMP), <span class="math inline">\(k7\)</span> (Sox9 effect on Wnt), and how much Bmp diffusion is faster than that of Wnt. Change one parameter at a time and realise that changing in one direction is sufficient; change in the other direction can be expected to have the opposite effect.</p>
<p>What effects on pattern wavelength do you find?</p>
<p><strong>Note</strong>: Make sure to use only small changes for <span class="math inline">\(k4\)</span> and <span class="math inline">\(k7\)</span>, if things start oscillating wildly or go blank you are watching numerical errors rather than real outcomes. When increasing Bmp diffusion you may need to reduce the time step for integration <span class="math inline">\(dt\)</span>.</p>
</div>
<blockquote class="blockquote">
<p>For simplicity I will only show Sox9 patterns:</p>
</blockquote>
<p><a href="images/sox9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="images/sox9.png" id="vegmini1" class="img-fluid" style="width:40.0%"></a></p>
<p><a href="images/sox9sweep.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="images/sox9sweep.png" class="img-fluid"></a></p>
<blockquote class="blockquote">
<p>In general, all three components follow similar trends in terms of concentrations as any impact on one component will affect the others via feedbacks.<br><br> Increasing or decreasing the parameters k2 and k4 has the same effect on the pattern. <br><br> With increased Sox9 activation (high k2) a lower amount of BMP is sufficient to get Sox9 to increase which in turn inhibits both Wnt and Bmp. As a result, the “peaks” are reached sooner in space resulting in shorter lower amplitude wavelengths. Similarly, when Sox9 inhibition of BMP is high (high k4) then BMP peaks are lower, which means there is less Sox9 activation leading to smaller amplitude peaks and hence shorter wavelength. Conversely, if Sox9 activation by BMP is weak (low k2), then it takes more BMP to build up to a peak, resulting in longer higher amplitude wavelengths. When Sox9 inhibition of BMP is weak (low k4), then it takes larger Sox9 domains to deplete BMP, resulting in longer higher amplitude wavelengths. If Sox9 activation (k2) or BMP inhibition (k4) are extremely strong, Sox9 locally depletes BMP faster than it can diffuse, which in turn leads to a reduction in Sox9; this could explain local oscillations and travelling waves, though this result should be verified with a better numerical scheme. Changing BMP self-inhibition (k5) has the opposite effect as Sox9 activation (k2) and BMP inhibition (k4). BMP self-inhibition restricts the size of BMP domains. Stronger self-inhibition (high k5) means small pockets of BMP are destabilized and Sox9 regions grow until the point that they completely deplete the local BMP, resulting in larger higher amplitude peaks. Vice-versa, reducing BMP self-inhibition (low k5) allows BMP peaks to more easily form, leading to shorter pattern wavelengths. Unlike the other two parameters, this reduction only happens up to a point, as the other feedback interactions take over.</p>
</blockquote>
<p><a href="images/sox9sweep2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="images/sox9sweep2.png" class="img-fluid"></a></p>
<blockquote class="blockquote">
<p>Increasing or decreasing the parameters k3 and k7 has the same effect on the pattern. When Wnt inhibition of Sox9 is increased (high k3) or Sox9 inhibition of Wnt is increased (high k7) the cross-inhibition of both variables is stronger, meaning they will tend to create larger exclusion zones which translates as longer higher amplitude wavelengths. When Wnt inhibition of Sox9 is decreased (low k3) or Sox9 inhibition of Wnt is decreased (low k7) smaller regions can coexist, reducing the wavelength and the amplitude. Even lower cross-inhibition means the regions overlap, and small time delays could lead to oscillations, although this should be verified with a numerically better scheme. Wnt self-inhibition (k9) limits the growth of Wnt domains. When self-inhibition is strong (high k9) Wnt domains shrink and therefore the wavelength is shorter with lower amplitude. Even stronger self-inhibition leads to overlapping domains and similar oscillations as observed for low k3 and low k7. Vice-versa, a decrease of Wnt self-inhibition (low k9) allows larger patches of Wnt to form, which increases the wavelength and the amplitude.</p>
</blockquote>
<p><a href="images/sox9sweep3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="images/sox9sweep3.png" class="img-fluid"></a></p>
<blockquote class="blockquote">
<p>The equation for Sox9 includes a “saturation term” in the form of a negative cubic power of Sox9, which means that effectively there is Sox9 self-inhibition in the model. The parameter delta multiplies a quadratic Sox9-dependent term in the equation for Sox9, which by default is set to 0. A positive delta effectively means that Sox9 has a self-activation term that counteracts the self-inhibition by the cubic saturation term, whereas a negative delta means the Sox9 self-inhibition is stronger. Introducing a stronger Sox9 self-activation (delta &gt; 0) increases the amplitude of Sox9 domains. Due to this local strong increase in Sox9, there is also a local strong depletion of BMP which prevents stripes from elongating, resulting in a spot pattern. Introducing a stronger Sox9 self-inhibition (delta &lt; 0) reduces the amplitude of Sox9 domains. There is thus less BMP depletion which enables stripes to elongate and fuse more easily.<br><br> Wnt and Sox9 are cross-inhibitory, and thus their domains will compete. Lower Wnt diffusion means Wnt domains form locally, leading to a reduced wavelength. Vice-versa a higher Wnt diffusion means that Wnt spreads away from Sox9 domains enabling them to expand ultimately resulting in longer wavelengths. Note that since BMP diffusion scales with Wnt diffusion, both parameters were changed at the same time.<br><br> The ratio of diffusion speeds defines the size of the Turing space. As the diffusion of BMP is set equal to that of Wnt, the homogenous solution becomes stable. Reducing the diffusion of BMP further leads to travelling waves or oscillations. This is because a slow (or stationary) BMP is in a local negative feedback loop with Sox9, which results in oscillatory behavior. The establishment of Wnt domains around the oscillating Sox9 regions is also oscillatory.<br><br> Increasing BMP diffusion allows to more effectively “feed” Sox9 domains, which leads to a subtle increase in the wavelength and amplitude.</p>
</blockquote>
<hr>
</section>
<section id="patterning-in-a-growing-domain-size" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="patterning-in-a-growing-domain-size"><span class="header-section-number">3.3</span> Patterning in a growing domain size</h2>
<p>Besides parameter values, the pattern that emerges from a Turing system is also strongly influenced by initial conditions and the size and shape of the domain where reactions and diffusion happen.</p>
<p>In the script <code>02__digits_growingsquare_PD.py</code> we added tissue growth in the proximo-distal (body to limb) direction. Growth is controlled by a growth rate <code>vi</code>.</p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.3 (Biology)</strong></span> &nbsp;</p>
<ul>
<li>Study the code to understand what it does.<br>
</li>
<li>Increase the horizontal growth rate <code>vi</code> in small increments (try: 0.01, 0.05, 0.1, 0.2, 0.5) and study what happens to the pattern. Explain why you think this happens.</li>
</ul>
<p><strong>Note:</strong> Increase the <code>totaltime</code> parameter to 5000 when using growth rate of 0.01.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> At low growth rates, the pattern forms stripes parallel to the growth direction. As the growth speed increases, the alignment shifts to stripes perpendicular to the growth direction. When growth is slow, the “new domain” added by growth is too small to fit a Turing wavelength. As a result, the existing pattern gets extended into the new parts. When growth is fast, the “new domain” added by growth is sufficiently large to fit a Turing wavelength, which becomes the selected mode. Additionally for faster growth, the diffusion/growth ratio is now lower making it harder to simple extend the existing pattern from diffusion of the old cells that previously made up the boundary.</p>
</blockquote>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="images/growth1d.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Simulations from this practical with domain growth"><img src="images/growth1d.png" class="img-fluid figure-img" alt="Simulations from this practical with domain growth"></a></p>
<figcaption>Simulations from this practical with domain growth</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="images/experimental_stripes.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="Experimental work using a light-gated chemical reaction that forms Turing patterns when illuminated. The illuminated area was shifted over time to produce a growing domain, work by @miguez2006effect"><img src="images/experimental_stripes.png" class="img-fluid figure-img" style="width:40.0%" alt="Experimental work using a light-gated chemical reaction that forms Turing patterns when illuminated. The illuminated area was shifted over time to produce a growing domain, work by Mı́guez et al. (2006)"></a></p>
<figcaption>Experimental work using a light-gated chemical reaction that forms Turing patterns when illuminated. The illuminated area was shifted over time to produce a growing domain, work by <span class="citation" data-cites="miguez2006effect">Mı́guez et al. (<a href="references.html#ref-miguez2006effect" role="doc-biblioref">2006</a>)</span></figcaption>
</figure>
</div>
<hr>
</section>
<section id="patterning-along-two-growth-axes" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="patterning-along-two-growth-axes"><span class="header-section-number">3.4</span> Patterning along two growth axes</h2>
<p>In reality, the limb bud grows in both the proximo-distal and anterior-posterior direction. In the script <code>03__digits_growingsquare.py</code> this is implemented with growth rates <code>vi</code> for proximo-distal growth and <code>vj</code> for anterior-posterior growth.</p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.4 (Biology)</strong></span> Play with the relative size of these growth rates and study what happens. Compare the order of appearance of stripes to the experimental data in the paper (see Figure 3F in the Raspopovic paper <span class="citation" data-cites="raspopovic2014digit">Raspopovic et al. (<a href="references.html#ref-raspopovic2014digit" role="doc-biblioref">2014</a>)</span>).</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> <img src="images/domaingrowth.png" id="vegmini1" class="img-fluid" style="width:140.0%" alt="Simulations from this practical with domain growth"> As in the previous question, the rate of growth influences whether stripes grow parallel to or perpendicular to the growth boundary. Also as before, slow horizontal growth promotes horizontal stripes (vi&lt;=0.1), and when this becomes faster stripes reorient to become vertical. However now, if we add vertical growth and it is not small enough relative to the horizontal growth, above and below the horizontal stripes we get vertical stripes and these domains are larger the faster vertical growth is relative to horizontal growth. So to get ok-ish patterns we need horizontal growth not too fast, and vertical growth to be slower. These results are consistent with horizontal growth of the “paddle” being faster than vertical growth. Also, we see that middle horizontal stripes appear earlier than the outer ones (in those situation where not all is vertical), similar to what we see in the paper in Fig 3F. Still interestingly, if you put vj=0.2 and vi=0.1 you promote initially vertical and later on horizontal stripes, this is an effect of in the vertical direction hitting the boundaries of the domain first and then still slowly continuing growth in the horizontal dimension at a not too high rate that was supportive of extending horizontal stripes.</p>
</blockquote>
<p><a href="images/horizontal.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="images/horizontal.png" id="vegmini1" class="img-fluid" style="width:40.0%"></a></p>
<hr>
</section>
<section id="making-a-virtual-tissue-grow" class="level2" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="making-a-virtual-tissue-grow"><span class="header-section-number">3.5</span> Making a virtual tissue grow</h2>
<p>Think about how tissue growth is implemented in the script <code>03__digits_growingsquare.py</code>.</p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.5 (Algorithmic and conceptual thinking)</strong></span> What is the default value of the newly added tissue when it grows? Is the concentration elsewhere in the tissue changed?</p>
<p>Do you think this approach is reasonable to model biological growth of a tissue? Explain why/why not.</p>
<p><em>Hint:</em> Think about it from the perspective of a growing cell containing a number of molecules “X”.<br>
If the cell increases in volume and does not produce/degrade X, what happens to the concentration of X?<br>
What would be the concentration of X in the two daughter cells if the cell divides?<br>
How would the situation change if X is constantly produced and degraded?</p>
<p>In the paper, @<span class="citation" data-cites="raspopovic2014digit">Raspopovic et al. (<a href="references.html#ref-raspopovic2014digit" role="doc-biblioref">2014</a>)</span> create a “tissue growth map” (see Figure 3A), which they use to map concentration values from one timepoint to the next using interpolation-based transformations.</p>
<p>Inspired by this approach, the script <code>04__digits_growingsquare.py</code> implements a different growth function that uses bilinear interpolation to expand the tissue. Play around with the growth rate parameters and compare the results to what you found earlier. Do the patterns differ? If yes, how do they differ?</p>
<p><em>Hint:</em> Try the supplementary script <code>supp_g__bilinear_interpolation.py</code> to understand how interpolation-based growth works.</p>
<p><strong>For master students:</strong> yet another way of implementing tissue growth would be to take cell division and inheritance of maternal state by the two daughter cells literal and implement this by new boundary cells copying the state of their direct neighbors. Implement this alternative growth and see how this affects your results.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong></p>
<ul>
<li>The default value of newly added tissue is 0, because the entire matrix is initialized with 0s, and only a small rectangle corresponding to the initial tissue domain is filled with random values and then updated based on the equations. When growth happens, the tissue domain extends to cover new parts of the matrix that were not included in the initialization. Extending the edges into 0-valued territory would be as if new, empty cells appear at the edges of the tissue, which is not a reasonable approach to model limb bud growth.<br><br></li>
<li>If a cell increases in volume without producing/degrading molecules “X”, then the concentration of X will decrease as the cell increases in volume. Daughter cells would have on average half the concentration of X as the mother cell. If “X” is constantly produced and degraded (and the cell is a well-mixed environment) then the concentration of X could be in quasi-steady state during growth. A growing cell would maintain the same concentration of X over time. When the cell divides, both daughter cells would inherit the same concentration of X as the mother cell had. Assuming the latter scenario is more likely, then growth of a tissue should work by “stretching” the values of the growing parts. (Some tissues actually do grow only at the edges, but even there we would expect the newly added “territory” to be made from previous cells and therefore still inherit their concentrations.)</li>
</ul>
</blockquote>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="images/growth1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11" title="Simulations from this practical with domain growth"><img src="images/growth1.png" class="img-fluid figure-img" style="width:140.0%" alt="Simulations from this practical with domain growth"></a></p>
<figcaption>Simulations from this practical with domain growth</figcaption>
</figure>
</div>
<blockquote class="blockquote">
<p>The different growth appears to affect pattern formation, e.g for vj=0.01 the pattern is still predominantly horizontal for 0.15, which was less so before. In constrast for vj=0.05 and vi=0.15 pattern is mostly vertical whereas in other growth regime it was still partly horizontal. Seems that transition from horizontal to vertical stripes is thus modulated. So for low vj it is easier to get horizontal stripes.</p>
</blockquote>
<hr>
</section>
<section id="spatial-modulation-of-parameters-k4-and-k7" class="level2" data-number="3.6">
<h2 data-number="3.6" class="anchored" data-anchor-id="spatial-modulation-of-parameters-k4-and-k7"><span class="header-section-number">3.6</span> Spatial modulation of parameters k4 and k7</h2>
<p>Let us now ignore growth and its role on orienting stripes for a while and explore once again the influence of parameters on the patterns. In reality, digits are patterned further apart at the distal end than at the proximal end, where they need to converge on a hand and wrist. This implies that the wavelength of the Turing pattern should not be constant. In a previous question you probably found that the <code>k7</code> and <code>k4</code> parameters impact the wavelength of the Turing pattern. The authors speculate that the FGF and Hox gene gradients observed in the limb bud exert an effect on the Sox9-BMP-Wnt patterning module through these parameters.</p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.6 (Biology)</strong></span> The script <code>05__digits_squarek4k7gradient.py</code> allows you to implement exponential gradients of <code>k4</code> and <code>k7</code> across the tissue. You can set the minimum and maximum values, axis (x=horizontal, y=vertical) and direction (0=decreasing, 1=increasing). Look in the supplement of the Raspopovic paper, page 48 Fig 2C (<span class="citation" data-cites="raspopovic2014digit">Raspopovic et al. (<a href="references.html#ref-raspopovic2014digit" role="doc-biblioref">2014</a>)</span>), left for the estimated change of <code>k4</code> and <code>k7</code> across the tissue and try to reproduce (not exactly but similarly) Figure 2C, right side, by playing with the parameters affecting the <code>k4</code> and <code>k7</code> gradients. Describe along which axis, in which direction and to what extent <code>k4</code> and <code>k7</code> are changing. Estimate the change in Turing pattern wavelength this results in.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> See the script <code>05__digits_squarek4k7gradient_answer.py</code> for a solution. The output may look sometimes like this: <img src="images/spatial_parameter_turing.png" class="img-fluid"></p>
<p>Parameter values for the gradients should be:</p>
<p><span class="math inline">\(k4_{min}=0.8\)</span>, <span class="math inline">\(k4_{max}=1.8\)</span>, <span class="math inline">\(direction=0\)</span>, <span class="math inline">\(axis ="x"\)</span></p>
<p><span class="math inline">\(k7_{min}=0.5\)</span>, <span class="math inline">\(k7_{max}=1.3\)</span>, <span class="math inline">\(direction=1\)</span>, <span class="math inline">\(axis="x"\)</span></p>
<p>This results in <span class="math inline">\(k4\)</span> decreasing and <span class="math inline">\(k7\)</span> increasing along the x-axis. Wavelength of the Turing pattern increases more than 2-fold</p>
</blockquote>
<hr>
</section>
<section id="hoxd13-and-fgf" class="level2" data-number="3.7">
<h2 data-number="3.7" class="anchored" data-anchor-id="hoxd13-and-fgf"><span class="header-section-number">3.7</span> Hoxd13 and FGF</h2>
<p>As mentioned above, the biological factors thought to underly the variation in the <code>k4</code> and <code>k7</code> parameters are Hoxd13 and FGF. Hoxd13 is highly expressed in the distal domain of the paddle and not expressed elsewhere, and this expression domain is growing as the paddle grows. FGF is expressed from the distal edge of the paddle and spreads through diffusion. For an illustration see Fig S8A of the supplement of the Raspopovic paper (page 10 <span class="citation" data-cites="raspopovic2014digit">Raspopovic et al. (<a href="references.html#ref-raspopovic2014digit" role="doc-biblioref">2014</a>)</span>). Experimental data furthermore suggest that Hoxd13 and FGF together affect <code>k4</code> and <code>k7</code>, which the authors decided to model using the following equations:</p>
<p><span class="math display">\[
k4^* = k4 - k_{HF_{bmp}} * fgf(i,j) * hox(i,j)
\]</span></p>
<p><span class="math display">\[
k7^* = k7 + k_{HF_{wnt}} * fgf(i,j) * hox(i,j)
\]</span></p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.7 (Algorithmic and biological thinking)</strong></span> The script <code>06__digits_squarehoxdfgf.py</code> implements how <code>k4</code> and <code>k7</code> are a function of local FGF and Hox values, but does not yet include Hox and FGF spatial patterns and how these develop over time. Adjust the code to incorporate the observed Hoxd13 and FGF patterns and study how the digit patterns compare to what you found under question 7. Note that arrays for <code>hox</code> and <code>fgf</code> are included, and decay and diffusion values for <code>fgf</code> are provided already in the code. Also note that <code>hox</code> and <code>fgf</code> each have a maximum value of 1.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> See the script <code>06_digits_squarehoxdfgf_answer.py</code> for a solution. The output may look sometimes like this:</p>
</blockquote>
<p><img src="images/spatial_parameter_turing2.png" class="img-fluid"> &gt; The Hox domain starts as a narrow vertical band and expands backward. The FGF gradient develops from the rightward boundary. Since a certain amount of Hox and FGF is needed to result in k4 and k7 values permissive of Turing patterns the stripes develop from right to left and because of the wavelength changes induced by different Hox and FGF values become closer together towards the left. Note that here a nice example run is shown but patterns are variable.</p>
<hr>
</section>
<section id="shape-makes-the-tissue" class="level2" data-number="3.8">
<h2 data-number="3.8" class="anchored" data-anchor-id="shape-makes-the-tissue"><span class="header-section-number">3.8</span> Shape makes the tissue</h2>
<p>Having studied the effects of growth and genetic factors modulating the Turing pattern now it is time for the final step. So far, we’ve approximated the limb bud very crudely with a rectangle. As you probably noticed in your parameter exploration, the tissue boundary can bias stripe orientation. Therefore it is interesting to study the interplay between the shape of the “paddle” or limb bud in which the digits develop and how it grows with the temporal development of the hox and fgf patterns. In the script <code>07__digits_growingpaddle.py</code> we implemented two different tissue geometries: An ellipse and a “paddle” that imitates the shape of the limb bud. Note that to implement Hox and FGF domains on a complex growing shape, quite a bit of complex code involving masks describing the tissue domain and angles to control the Hox and FGF domains was needed. At first, we are going to ignore these technical details.</p>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.8 (Biology)</strong></span> Switching between geometries is easily done by changing the value of the parameter <code>geometry</code> on line 78 of the script. Start with the “ellipse” geometry. What do you observe with regards to the FGF gradient? What consequences does this have? Why would this be “a smart thing to do”?</p>
<p>Play around with <code>vi/vj</code> and <code>Lx0</code> to investigate the effect of ellipse shape and development over time on the number, shape and robustness of digits that form.</p>
<p>Now switch to the paddle geometry. Observe the patterns that emerge and compare these to the simulated and experimentally observed patterns in Figure 3E and 3F of the Raspopovic article (<span class="citation" data-cites="raspopovic2014digit">Raspopovic et al. (<a href="references.html#ref-raspopovic2014digit" role="doc-biblioref">2014</a>)</span>). According to you what does and what doesn’t the model explain well?</p>
<p>To implement non-square geometries, the script uses “masks”. Study how these masks are created in the functions <code>create_ellipse_mask</code> and <code>create_tissue_mask</code>.</p>
<p><em>Hint:</em> Try the supplementary script <code>supp_i__geometries.py</code>.</p>
<p><em>Note:</em> For computational convenience, we use the “Class” data structure to implement common functions shared by the masks. This is a somewhat more advanced programming concept, so it’s ok if you don’t understand what it does yet. For the extra curious and motivated, feel free to check out the supplementary script <code>supp_i__class_data_structure.py</code>.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong><br></p>
<ul>
<li>Fgf gradient forms from bottom to top, as a consequence bottom fingers form first, often one of the bottom fingers forms and than splits. While in general Turing mechanisms give patterns they are not reproducible in terms of the positioning of the stripes. By starting the highest point of the FGF gradient locally it is determine where the first digits can arise, generating control over the positioning of the digits.<br><br></li>
<li>Vi 0.1 goes okay, 0.2 stretched and splitting of digits <br> Vj too low too little room so less than 5 digits, vi too high more than 5 fingers, higher rates result in less reproducible patterns Below a nice pattern showing 5 fingers, but also showing near splitting of one of the fingers at the end<br><br></li>
<li>The model explains the formation and positioning of the 4 digits, and how the distance between these digits increases along the limb bud. However, the model clearly does not simulate/explain the formation of the wrist bones that is observed experimentally. Interestingly the experimental data suggest that while digits may form from the “right to the left”, the “leftmost” bones of the wrist are there the earliest, suggesting there locally different conditions apply allowing earlier Turing patterning but with a very large wavelength happening there. <br><br></li>
</ul>
</blockquote>
<p><a href="images/fingers.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="images/fingers.png" class="img-fluid"></a></p>
<hr>
</section>
<section id="master-students-exercise" class="level2" data-number="3.9">
<h2 data-number="3.9" class="anchored" data-anchor-id="master-students-exercise"><span class="header-section-number">3.9</span> Master students exercise</h2>
<div id="exr-tur" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.9 (Poly and oligodactyly)</strong></span> Above we played with ellipse size and growth rates, however in addition to mutations affecting tissue size and growth, also mutations in genes affecting the Hox and FGF morphogens may occur. Play with parameters that determine Hox and FGF expression zones on the paddle geometry. Which parameter changes lead to the formation of supernumerary fingers (polydactyly)? Which parameters lead to too few fingers (oligodactyly)?</p>
<p><em>Hint:</em> The following supplementary scripts will be helpful to understand how the FGF and Hox domains were coded:</p>
<ul>
<li><code>supp_i__ellipse_slice.py</code> visualizes the usefulness of masks based on the example of selecting a slice of an ellipse;</li>
<li><code>supp_i__smooth_function.py</code> visualizes the function <code>smooth_function()</code>;</li>
<li><code>supp_i__morphological_operations.py</code> visualizes morphological operations (binary dilate/erode and Gaussian blur) which are used for mask operations, calculating the Laplacian on the irregular domain, and to select pixels to generate the FGF and Hox pattern;</li>
<li><code>supp_i__maskDifference.py</code> visualizes the difference of masks, which is how the function <code>single_step_growth()</code> checks if the tissue domain is growing or shrinking.</li>
</ul>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> Reference values: <img src="images/referencevalues.png" id="vegmini1" class="img-fluid" style="width:140.0%"></p>
</blockquote>
<blockquote class="blockquote">
<p>Decreasing FGF diffusion increases the number of fingers and creates a fin-like appearance. However note that it also creates a shorter finger forming domain. Increasing FGF diffusion decreases the number of fingers.</p>
</blockquote>
<p><img src="images/polydact.png" class="img-fluid"> <img src="images/growthanswer.png" class="img-fluid"></p>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<ol type="1">
<li><span class="citation" data-cites="turing1990chemical">Turing (<a href="references.html#ref-turing1990chemical" role="doc-biblioref">1990</a>)</span></li>
<li><span class="citation" data-cites="raspopovic2014digit">Raspopovic et al. (<a href="references.html#ref-raspopovic2014digit" role="doc-biblioref">2014</a>)</span></li>
<li><span class="citation" data-cites="miguez2006effect">Mı́guez et al. (<a href="references.html#ref-miguez2006effect" role="doc-biblioref">2006</a>)</span></li>
</ol>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-miguez2006effect" class="csl-entry" role="listitem">
Mı́guez, David G, Milos Dolnik, Alberto P Munuzuri, and Lorenz Kramer. 2006. <span>“Effect of Axial Growth on Turing Pattern Formation.”</span> <em>Physical Review Letters</em> 96 (4): 048304.
</div>
<div id="ref-raspopovic2014digit" class="csl-entry" role="listitem">
Raspopovic, Jelena, Luciano Marcon, Laura Russo, and James Sharpe. 2014. <span>“Digit Patterning Is Controlled by a Bmp-Sox9-Wnt Turing Network Modulated by Morphogen Gradients.”</span> <em>Science</em> 345 (6196): 566–70.
</div>
<div id="ref-turing1990chemical" class="csl-entry" role="listitem">
Turing, Alan Mathison. 1990. <span>“The Chemical Basis of Morphogenesis.”</span> <em>Bulletin of Mathematical Biology</em> 52 (1): 153–97.
</div>
</div>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./pattern_practical_1.html" class="pagination-link" aria-label="Gradients and segments">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Gradients and segments</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./pattern_practical_3.html" class="pagination-link" aria-label="Clock-and-wavefront patterning">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Clock-and-wavefront patterning</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Theoretical Biology, Utrecht University.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was made with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"loop":false,"openEffect":"zoom","selector":".lightbox","closeEffect":"zoom","descPosition":"bottom"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>