<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>4&nbsp; Clock-and-wavefront patterning – Modeling Life</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./morpho_intro_text.html" rel="next">
<link href="./pattern_practical_2.html" rel="prev">
<link href="./images/favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./pattern_intro_text.html">I) Pattern formation</a></li><li class="breadcrumb-item"><a href="./pattern_practical_3.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Clock-and-wavefront patterning</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./images/android-chrome-512x512.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Modelling life</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false">
 <span class="menu-text">Course information</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Modeling life</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./general.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">General course info</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slides</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro_to_python.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">0) Introduction to Python</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">I) Pattern formation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Pattern formation (intro)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_practical_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Gradients and segments</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_practical_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Turing digit patterns</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pattern_practical_3.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Clock-and-wavefront patterning</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false">
 <span class="menu-text">II) Morphogenesis</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./morpho_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">What is morphogenesis?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./morpho_cpm_practical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Erika CPM practical</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">III) Cell differentiation</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differentiation_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Differentiation introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differentiation_practical_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Gene regulation in time</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differentiation_practical_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Gene regulation in space</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false">
 <span class="menu-text">IV) Environment</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evdev_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Environmental inputs</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evdev_practical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Modeling Thermomorphogenesis in Plants</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false">
 <span class="menu-text">V) Evolution</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_intro_text.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to evolution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_practical_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Sticking together</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_practical_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">What gets selected?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evo_practical_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Public goods</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false">
 <span class="menu-text">Mini-projects</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Miniproject: Pattern formation in vegetation in semi-arid areas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Miniproject: maintaining a healthy microbiome</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Miniproject: The devil is in the details – auxin transport in the root</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Miniproject: Patterning – Asymmetric localization of ZIF-1 in the C. elegans embryo</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Miniproject: Oncogenic propensity of different stem cell niches</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="-1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul class="collapse">
  <li><a href="#goal-of-the-tutorial" id="toc-goal-of-the-tutorial" class="nav-link active" data-scroll-target="#goal-of-the-tutorial"><span class="header-section-number">4.1</span> Goal of the tutorial:</a></li>
  <li><a href="#the-model-system" id="toc-the-model-system" class="nav-link" data-scroll-target="#the-model-system"><span class="header-section-number">4.2</span> The model system</a></li>
  <li><a href="#programming-with-classes" id="toc-programming-with-classes" class="nav-link" data-scroll-target="#programming-with-classes"><span class="header-section-number">4.3</span> Programming with classes</a></li>
  <li><a href="#questions" id="toc-questions" class="nav-link" data-scroll-target="#questions"><span class="header-section-number">4.4</span> Questions</a></li>
  <li><a href="#relevant-literature" id="toc-relevant-literature" class="nav-link" data-scroll-target="#relevant-literature"><span class="header-section-number">4.5</span> Relevant literature</a></li>
  <li><a href="#references-for-biological-parameters" id="toc-references-for-biological-parameters" class="nav-link" data-scroll-target="#references-for-biological-parameters"><span class="header-section-number">4.6</span> References for biological parameters:</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./pattern_intro_text.html">I) Pattern formation</a></li><li class="breadcrumb-item"><a href="./pattern_practical_3.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Clock-and-wavefront patterning</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Clock-and-wavefront patterning</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="goal-of-the-tutorial" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="goal-of-the-tutorial"><span class="header-section-number">4.1</span> Goal of the tutorial:</h2>
<p>In this tutorial you will look at gradient formation and patterning using different mechanisms than you have seen previously. We will model the so-called clock-and-wavefront pattern, which stems from oscillations in gene expression and gene products and results in a regularly striped segmentation pattern in a growing tissue. Note that as compared to the Drosophila French Flag type case, this is thought to be the evolutionary ancestral model of segmenting animal body axes.</p>
</section>
<section id="the-model-system" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="the-model-system"><span class="header-section-number">4.2</span> The model system</h2>
<p>The clock-and-wavefront model is an important model in describing somitogenesis. In this process early in embryo development, the somites, a precursor tissue for the vertebrae and other tissues later in development, are formed from the pre-somitic mesoderm. This mesodorm extends on the posterior end by growth, and the somites bud off periodically at the anterior end in the order of a couple of weeks (in humans).</p>
<p>Key players in this model system are the protein FGF (fibroblast growth factor), and many genes and gene products that have an oscillatory pattern that will determine cell fate. However, for this tutorial, we simplify this to a single pair of gene mRNA and product, denoted with <span class="math inline">\(m\)</span> and <span class="math inline">\(p\)</span> for short.</p>
</section>
<section id="programming-with-classes" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="programming-with-classes"><span class="header-section-number">4.3</span> Programming with classes</h2>
<p>Because we are going to make a more complex model with a tissue existing of multiple cells, and each cell having its own concentrations of FGF, <span class="math inline">\(m\)</span> and <span class="math inline">\(p\)</span>, we are going to use Classes in our code. You have been using Classes already: the data types such as <code>int</code>, <code>str</code> and <code>bool</code> have their own class, and the <code>str</code> class has many <em>methods</em> (=functions working on that class) defined, such as <code>"hello world".upper()</code>, but it is also possible to create custom classes. With classes, you can easily make objects, which is part of the object-oriented programming paradigm.</p>
<p>Today, we are going to use classes for the different levels of our model: 1) tissue, 2) cell, 3) <span class="math inline">\(m\)</span> &amp; <span class="math inline">\(p\)</span> clock and 4) the plotting. By using classes, we can separate things that happen on a tissue/cellular/clock scale, and seperate the model from the visualization of it. You will first work with 1 &amp; 2 &amp; 4, then 3 on its own and then combine all four yourself into one model.</p>
<p>Important concepts when working with classes are</p>
<ul>
<li>Class versus instance</li>
<li>Defining the <code>__init__</code> method and other methods</li>
<li>Class attributes and using the keyword <code>self</code></li>
</ul>
<p>This tutorial should be doable without an extensive knowledge of classes as there are plenty of examples to copy-paste from, but feel free to read up on these concepts here at <a href="https://www.datacamp.com/tutorial/python-oop-tutorial">this online tutorial</a></p>
</section>
<section id="questions" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="questions"><span class="header-section-number">4.4</span> Questions</h2>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.1 (Biology - An alternative gradient forming mechanism)</strong></span> In practical 1 we saw how gradients can be created through local production, diffusion and decay. However, other mechanisms for gradient formation are possible, such as cell lineage transport. Here we work with a model for the FGF gradient (<code>fgfgradientfromgrowth.py</code>) where only the rightmost/posterior cell produces FGF and grows, and in which cells upon division inherit this FGF from their mother cell. First study the code to see how it uses Classes and see what happens. Next, play with the model by varying the decay rate of FGF and the division rate of the cells. How does this affect the gradient? What happens if divisions are only allowed during the first half of the simulation (put <code>divisiontime</code> to 0.5 instead of 1).</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> If growth rate goes up cell volumes increase faster, causing more dilution of FGF and hence a lower maximum of the gradient as due to the stable protein quite some time is needed to compensate for this decrease by more production, at the same time the gradient is less steep as cell division follow up faster and hence less time has passed and less day has taken place.</p>
<p>If decay rate goes up, maximum goes down but the gradient become steeper and shorter as for same time between divisions more decay takes place. If divisions stop halfway the simulation FGF in last cell goes up (no longer any loss due to growth-induced dilution) and FGF in all other cells goes to zero (no fresh influx from newly divided cells still having FGF).</p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.2 (Conceptual thinking)</strong></span> When would this type of gradient formation be more applicable than the earlier studied production, diffusion, decay type of gradient formation? Compare how this model is built-up to a production/diffusion/decay gradient formation model.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> Earlier gradient model: gradient length increases and slope decreases with faster diffusion and slower degradation. Here gradient driven by combination of growth/division and degradation. This new mechanism can only work if growth and patterning are occurring simultaneously, not if growth precedes patterning..</p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.3 (Mathematics)</strong></span> Let us now move to the other half of the clock and wavefront model, the clock part (<code>clock.py</code>), in which we implemented one of the earliest models for the somitogenesis oscillator from <a href="https://doi.org/10.1016/S0960-9822(03)00534-7">Lewis (2003)</a> which models a gene that codes for a mRNA (<span class="math inline">\(m\)</span>) that encodes a protein (<span class="math inline">\(p\)</span>) that acts as a repressive transcription factor of this same gene. In class we discussed how for oscillations negative feedbacks, delays and non-linearity are important. Examine the code to find the differential equations governing this model and determine the negative feedback, delay and non-linearities in them.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> Non-linearity is in saturation function with power n, negative feedback is because affects own expression negatively, delay is here modeled explicitly through a special delay type differential equation</p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.4 (Biology)</strong></span> Play with the parameters of the model. How does the delay (<span class="math inline">\(\tau\)</span>) affect oscillations?</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> Larger value of <span class="math inline">\(\tau\)</span>=delay results in a longer period (lower frequency) and higher amplitude. Remember how in class we discussed how removing introns reduced delays and affected oscillator period.</p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.5 (Algorithmic thinking)</strong></span> In the file <code>rolling_clock.py</code>, there is a different implementation of the clock. Compare the two files and find out how they differ. What benefits for studying the model does <code>clock.py</code> have over <code>rolling_clock.py</code> and vice versa? Ignore the added functions <code>__copy__</code> and <code>set_tau</code> in this comparison. Some differences become clearer when you run the code too.</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> Working with delay functions implies that you need the value of a variable not on the current time but on time-<span class="math inline">\(\tau\)</span>. In the previous model we simply memorized the entire history of both variables. Here instead we only store the variable state for the length of the time delay <span class="math inline">\(\tau\)</span>. This saves memory and hence runtime. This will become especially important when simulating a larger number of individual cells each having their own clocks. In the plotting we only plotted stored values, hence you only see what happened between current_time-<span class="math inline">\(\tau\)</span> and current_time instead of what happened from start of simulation till current time. Of course if you want to dynamically change <span class="math inline">\(\tau\)</span> you need to make sure you store values over an interval corresponding to the largest possible value of <span class="math inline">\(\tau\)</span> and need to know this beforehand.</p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.6</strong></span> Next, we will combine the clock and FGF wavefront into a single model.</p>
<p><strong>Read and interpret existing code:</strong> Read the code of <code>clockplusfgf.py</code> and try to understand the assumptions from this model implementation by answering the following questions:</p>
<ol type="1">
<li>Each cell has its own clock. What clock states is the tissue initialized with? And what clock states do newly divided cells get? 2. The FGF wavefront affects <span class="math inline">\(\tau\)</span> : what function is used for that? What are your expectations for the effect of the FGF wavefront on the cells’ clocks? 3. Does growth affect the clock state? 4. What is your opinion on the assumptions discussed in the three questions?</li>
</ol>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong></p>
<ol type="1">
<li><p>Going to code rolling_clock.py we see that clocks are initialized with the <strong>init</strong> function of the Class Clock, which gets as arguments <span class="math inline">\(m0\)</span> and <span class="math inline">\(p0\)</span>, this initialization function is called when in Class <code>Cell</code>, in its respective <code>__init__</code> function <code>self.clock=Clock(..)</code> is called. It uses default values of <span class="math inline">\(m=0\)</span> and <span class="math inline">\(p=0\)</span> to pass to the <code>Clock</code>. (In turn, the initialization function of Class <code>Cell</code> is called from the initialize_regular_tissue function of Class <code>Tissue</code> when the first cell is created). Newly divided cells inherit clock state and and FGF value from mother cell.</p></li>
<li><p>The function is: <span class="math display">\[\tau_\text{cell}=\tau_\text{model}(1+0.5\frac{100-\text{FGF}\_\text{cell}}{100}),\]</span> So if localfgf=max=100, <span class="math inline">\(\tau{local}\)</span>=<span class="math inline">\(\tau{global}\)</span>, and if localfgf=min=0 <span class="math inline">\(\tau{local}\)</span>=1.5*<span class="math inline">\(\tau{global}\)</span> So for lower FGF <span class="math inline">\(\tau\)</span> increases slowing down the oscillations.</p></li>
<li><p>Not directly, growth of cells, leading to volume increase and potential dilution effects is not affecting m or p levels, however since it is affecting FGF levels it is indirectly affecting the clock.</p></li>
<li><p>Initialization is somewhat arbitrary, inheritance from mother cells is logical although in reality it is of course not perfect but noisy, FGF effect is also reasonable to get the observed slowing and waves, but it would be more logical for clock state like fgf to also be diluted if cells grow their volume</p></li>
</ol>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.7 (Biology)</strong></span> Describe how the model behaves and why. Do we get stable somites?</p>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> First there is a startup period where we have synchronized oscillations in the first formed cells that all still have high FGF. Then as the tissue grows and the FGF gradient forms we obtain waves of oscillations moving from right to left with increasing amplitudes. The right to left movement arises because oscillations are faster on the right than on the left, this difference in oscillation frequency arises from the impact of FGF on oscillator frequency. However, the oscillations do not become halted and transformed into a stable spatial pattern because there is no memory mechanism in place.</p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.8 (Biology &amp; Programming)</strong></span> What is still missing is a means to transform the temporal oscillations in the posterior of the tissue into a spatial pattern in the anterior. In the French Flag morphogen gradient lecture we discussed that memory mechanisms are important to stabilize spatial patterns once the start up signal that broke the symmetry and initialized patterning has gone. It turns out that such a memorization/stabilization mechanism is also essential to convert oscillations to stripes. Add a memorization mechanism to the model to achieve this. <em>Hint</em>: Make use of an extra ‘memory’ molecule <span class="math inline">\(M\)</span> in each cell and perform its updating inside the function <code>run_clocks</code>. To see what you are doing, visualize the spatial pattern of <span class="math inline">\(M\)</span><code>over time following the same procedure as for visualizing the</code>fgf` and <span class="math inline">\(p\)</span> values. Think about where, when and what should be memorized to design how <span class="math inline">\(M\)</span> is regulated and what <span class="math inline">\(M\)</span> itself affects and take a stepwise approach.</p>
<ol type="1">
<li><p>Add an extra ‘memory’ molecule <span class="math inline">\(M\)</span> as a new attribute to the Class <code>Cell</code>, which value is inherited from the mother cell upon division. To create memory we need bistability, which can be easily achieved by having the memory molecule having a non-linear saturating positive feedback on itself. However, doing only this it will depend on the initial value of <span class="math inline">\(M\)</span> -below or above a threshold- whether <span class="math inline">\(M\)</span> will autoactivate or not. Additionally, if it happens it will occur across the entire tissue.</p></li>
<li><p>Thus, we need to make <span class="math inline">\(M\)</span> dependent on the clock. To achieve this, we can start with low values of <span class="math inline">\(M\)</span> that prevent autoactivation, and then have <span class="math inline">\(M\)</span> activated by either <span class="math inline">\(p\)</span> (part of the clock) or <span class="math inline">\(M\)</span> itself using the following function: <span class="math display">\[\frac{dM}{dt}=c\max\left(\frac{p^4}{h_p^4+p^4}, \frac{M^4}{h_M^4+M^4}\right)-\delta M.\]</span></p>
<p>However, now we simply always get activation of <span class="math inline">\(M\)</span> everywhere.</p></li>
<li><p>Since oscillations occur in the PSM and stripe formation occurs only more anteriorly we should constrain memorization to occur only below a certain FGF value. As a consequence, high levels of <span class="math inline">\(M\)</span> now only arise anteriorly, but there is no pattern yet and oscillations keep occurring. This is because cells cycle through high and low p values so all cells at some point have high <span class="math inline">\(p\)</span> values and can induce <span class="math inline">\(M\)</span>.</p></li>
<li><p>So in addition to have memorization occur only below a certain FGF value it should also occur above another, lower FGF value, constraining it to occur in a limited temporal window that enables cells passing through there with a high <span class="math inline">\(p\)</span> state to induce a high <span class="math inline">\(M\)</span> state while cells pasing through with a low <span class="math inline">\(p\)</span> state to not induce a high <span class="math inline">\(M\)</span> state. You should get somite patterns that are stable, but not necessarily regular. Still <span class="math inline">\(m\)</span> and <span class="math inline">\(p\)</span> oscillations continue.</p></li>
<li><p>As a final step, beyond the FGF window where we memorize <span class="math inline">\(M\)</span> we no longer update either <span class="math inline">\(M\)</span> or the clock.</p></li>
</ol>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> See code <code>05_answers_clockplusfgf_memory.py</code></p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.9 (Biology)</strong></span> Both zebrafish and mice are common model organisms, so we know a lot about the biological parameters of their somitogenesis. See the following list:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 28%">
<col style="width: 11%">
<col style="width: 18%">
<col style="width: 25%">
<col style="width: 16%">
</colgroup>
<thead>
<tr class="header">
<th>Parameter</th>
<th>Zebrafish</th>
<th>Reference</th>
<th>Mouse</th>
<th>Reference</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Duration of somitogenesis</td>
<td>18 h</td>
<td>(a)</td>
<td>5 days</td>
<td>(b)</td>
</tr>
<tr class="even">
<td>Number of somites</td>
<td>~30</td>
<td>(a)</td>
<td>65</td>
<td>(c)</td>
</tr>
<tr class="odd">
<td>Somite size</td>
<td>50 <span class="math inline">\(\mu\)</span> / 30 <span class="math inline">\(\mu\)</span></td>
<td>(d) (a)</td>
<td>120 <span class="math inline">\(\mu\)</span></td>
<td>(c)</td>
</tr>
<tr class="even">
<td>Cells per somite</td>
<td>~5 cell in diameter</td>
<td>(a)</td>
<td>5-11 (estimated from total cell size in 3D, ranging 1 order of magnitude)</td>
<td>(c)</td>
</tr>
<tr class="odd">
<td>Clock period</td>
<td>25 min /30 min</td>
<td>(d) / (e) / (f)</td>
<td>2-3 h</td>
<td>(e)</td>
</tr>
</tbody>
</table>
<p>From these parameters, you can derive a number of desired model inputs/outcomes:</p>
<ul>
<li>The total size of the tissue at the end of somitogenesis</li>
<li>The size of a cell</li>
<li>Speed of division</li>
</ul>
<p>Use the model and try to find suitable model parameters to recreate the development of both zebrafish and mice: is this model able to describe both of these processes? I.e., is this model able to deal with the scale differences between zebrafish and mice?</p>
<p>A couple of notes:</p>
<ul>
<li>You don’t have to exactly recreate the biological parameter with 100% precision, except for the number of segments/somites, although it can be a fun challenge to get a complete match.</li>
<li>You might want to adapt your plotting timestep to have sufficient but not too many plot updates in one simulation.</li>
<li>You can test the clock parameters separately with the <code>clock.py</code>/<code>rolling_clock.py</code> script. Don’t forget that FGF has an effect on <span class="math inline">\(\tau\)</span> !</li>
</ul>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> See codes <code>06_answers_clockplusfgf_memory_mouse.py</code> and <code>06_answers_clockplusfgf_memory_zebrafish.py</code>.</p>
<p><strong>Zebrafish:</strong></p>
<p>Totaltime: 18 hours for duration of somitogenesis + 10 hours startup phase.</p>
<p>Cell width= 30-50 <span class="math inline">\(\mu\)</span> sized somite/5 cells per somite = 6-10 so 8</p>
<p>Growthrate: 30 somites x 5 cells x 8<span class="math inline">\(\mu\)</span> / 18 hours = <span class="math inline">\(1200\)</span><span class="math inline">\(/18*3600\)</span>s</p>
<p>For clock frequency:</p>
<p><span class="math inline">\(\tau\)</span> times 0.25</p>
<p>Fac = 6 (multiplies alpha, beta, mu, v)</p>
<p>This gives 32/31 somites with on average 4.9 cells, so perfect.</p>
<p><strong>Mouse:</strong></p>
<p>Simply doing the computations would give you:</p>
<p>Totaltime: 5 days for duration of somitogenesis + 15 hours startup phase</p>
<p>Cell width = 120<span class="math inline">\(\mu\)</span> sized somite/5-11 cells = <span class="math inline">\(120/8=15\)</span></p>
<p>Growthrate: 65 somites x 8 cells x 15<span class="math inline">\(\mu\)</span> / 5 days = <span class="math inline">\(7800/5* 24 *3600\)</span>s</p>
<p>For clock frequency:</p>
<p><span class="math inline">\(\tau\)</span> times 1.4-1.6 is sufficient (Fac=1 (multiplies alpha, beta, mu, v))</p>
<p>Somehow this does not fully work:</p>
<p>Too few cells (not <span class="math inline">\(65*8=520\)</span>) too short tissue (not <span class="math inline">\(65*8*15=7800\)</span>) and only 24 segments. First, remember <span class="math inline">\(\tau_{local}=(1+0.5 *(100-localfgf)/100))*\tau_{global}\)</span>. So <span class="math inline">\(\tau_{local}\)</span> max 1.5 times <span class="math inline">\(\tau_{global}\)</span> so the also increasing <span class="math inline">\(\tau_{global}\)</span> 1.4-1.6 times 1.5 makes it really slow. Indeed <span class="math inline">\(\tau\)</span> times 1.2 gives 31 segments, <span class="math inline">\(\tau\)</span> times 1 gives &gt; 36 segments, <span class="math inline">\(\tau\)</span> times 0.7 gives 49 segments</p>
<p>Good to realize that the numbers in the table do not completely add up:</p>
<p>If it takes 5 days =<span class="math inline">\(5*24=120\)</span> hours, and the period is 2-3 hours this would result in 40-60 segments but never 65. For a duration of 5.5 days and <span class="math inline">\(\tau\)</span> times 0.6 we get to 62 somites of size 9, so quite close.</p>
</blockquote>
<div id="exr-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.10 (Questions for master students)</strong></span> The model currently has a number of assumptions that we can question. Feel free to explore any of these further open questions and study how it effects the outcome of the model:</p>
<ul>
<li>What if the relationship between FGF and <span class="math inline">\(\tau\)</span> is shaped differently? For instance, if the clock runs faster on the left/anterior than on the right/posterior?</li>
<li>What if the clock state of a daughter cell is started fresh rather than being a copy from the mother cell?</li>
<li>What if the clock <span class="math inline">\(m\)</span> and <span class="math inline">\(p\)</span> are diluted by growth?</li>
<li>What if <span class="math inline">\(\tau\)</span> is unaffected by FGF: can we still get somites fixed in place?</li>
</ul>
</div>
<blockquote class="blockquote">
<p><strong>Answer</strong> These are open questions and do not have a single correct answer. Feel free to discuss with us if you find something interesting or have an interesting answer.</p>
</blockquote>
</section>
<section id="relevant-literature" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="relevant-literature"><span class="header-section-number">4.5</span> Relevant literature</h2>
<p>If you want to know more about the model system and previous models, have a look at the following (after the tutorial):</p>
<p><span class="citation" data-cites="lewis2003autoinhibition">Lewis (<a href="references.html#ref-lewis2003autoinhibition" role="doc-biblioref">2003</a>)</span> Autoinhibition with transcriptional delay: a simple mechanism for the zebrafish somitogenesis oscillator.</p>
<p><span class="citation" data-cites="hester2012multi">Hester (<a href="references.html#ref-hester2012multi" role="doc-biblioref">2012</a>)</span> A Multi-cell, Multi-scale Model of Vertebrate Segmentation and Somite Formation.</p>
<p><span class="citation" data-cites="herrgen2010intercellular">Herrgen et al. (<a href="references.html#ref-herrgen2010intercellular" role="doc-biblioref">2010</a>)</span> Intercellular coupling regulates the period of the segmentation clock.</p>
<p><span class="citation" data-cites="soroldoni2014doppler">Soroldoni et al. (<a href="references.html#ref-soroldoni2014doppler" role="doc-biblioref">2014</a>)</span> Genetic oscillations. A Doppler effect in embryonic pattern formation.</p>
<p><span class="citation" data-cites="sonnen2018modulation">Sonnen et al. (<a href="references.html#ref-sonnen2018modulation" role="doc-biblioref">2018</a>)</span> Modulation of Phase Shift between Wnt and Notch Signaling Oscillations Controls Mesoderm Segmentation.</p>
<p><span class="citation" data-cites="bulusu2017spatiotemporal">Bulusu et al. (<a href="references.html#ref-bulusu2017spatiotemporal" role="doc-biblioref">2017</a>)</span> Spatiotemporal Analysis of a Glycolytic Activity Gradient Linked to Mouse Embryo Mesoderm Development.</p>
<p><span class="citation" data-cites="van2025scaling">Oostrom et al. (<a href="references.html#ref-van2025scaling" role="doc-biblioref">2025</a>)</span> Coupling of cell proliferation to the segmentation clock ensures robust somite scaling.</p>
</section>
<section id="references-for-biological-parameters" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="references-for-biological-parameters"><span class="header-section-number">4.6</span> References for biological parameters:</h2>
<ol type="a">
<li><p><span class="citation" data-cites="stickney2000somite">Stickney, Barresi, and Devoto (<a href="references.html#ref-stickney2000somite" role="doc-biblioref">2000</a>)</span></p></li>
<li><p><span class="citation" data-cites="saga2012mechanism">Saga (<a href="references.html#ref-saga2012mechanism" role="doc-biblioref">2012</a>)</span></p></li>
<li><p><span class="citation" data-cites="tam1981control">Tam (<a href="references.html#ref-tam1981control" role="doc-biblioref">1981</a>)</span></p></li>
<li><p><span class="citation" data-cites="ishimatsu2018size">Ishimatsu et al. (<a href="references.html#ref-ishimatsu2018size" role="doc-biblioref">2018</a>)</span></p></li>
<li><p><span class="citation" data-cites="carraco2022vertebrate">Carraco, Martins-Jesus, and Andrade (<a href="references.html#ref-carraco2022vertebrate" role="doc-biblioref">2022</a>)</span></p></li>
<li><p><span class="citation" data-cites="tomka2018travelling">Tomka, Iber, and Boareto (<a href="references.html#ref-tomka2018travelling" role="doc-biblioref">2018</a>)</span></p></li>
</ol>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-bulusu2017spatiotemporal" class="csl-entry" role="listitem">
Bulusu, Vinay, Nicole Prior, Marteinn T Snaebjornsson, Andreas Kuehne, Katharina F Sonnen, Jana Kress, Frank Stein, Carsten Schultz, Uwe Sauer, and Alexander Aulehla. 2017. <span>“Spatiotemporal Analysis of a Glycolytic Activity Gradient Linked to Mouse Embryo Mesoderm Development.”</span> <em>Developmental Cell</em> 40 (4): 331–41.
</div>
<div id="ref-carraco2022vertebrate" class="csl-entry" role="listitem">
Carraco, Gil, Ana P Martins-Jesus, and Raquel P Andrade. 2022. <span>“The Vertebrate Embryo Clock: Common Players Dancing to a Different Beat.”</span> <em>Frontiers in Cell and Developmental Biology</em> 10: 944016.
</div>
<div id="ref-herrgen2010intercellular" class="csl-entry" role="listitem">
Herrgen, Leah, Saúl Ares, Luis G Morelli, Christian Schröter, Frank Jülicher, and Andrew C Oates. 2010. <span>“Intercellular Coupling Regulates the Period of the Segmentation Clock.”</span> <em>Current Biology</em> 20 (14): 1244–53.
</div>
<div id="ref-hester2012multi" class="csl-entry" role="listitem">
Hester, Susan D. 2012. <span>“Multi-Scale Cell-Based Computational Models of Vertebrate Segmentation and Somitogenesis Illuminate Coordination of Developmental Mechanisms Across Scales.”</span> PhD thesis, Indiana University.
</div>
<div id="ref-ishimatsu2018size" class="csl-entry" role="listitem">
Ishimatsu, Kana, Tom W Hiscock, Zach M Collins, Dini Wahyu Kartika Sari, Kenny Lischer, David L Richmond, Yasumasa Bessho, Takaaki Matsui, and Sean G Megason. 2018. <span>“Size-Reduced Embryos Reveal a Gradient Scaling-Based Mechanism for Zebrafish Somite Formation.”</span> <em>Development</em> 145 (11): dev161257.
</div>
<div id="ref-lewis2003autoinhibition" class="csl-entry" role="listitem">
Lewis, Julian. 2003. <span>“Autoinhibition with Transcriptional Delay: A Simple Mechanism for the Zebrafish Somitogenesis Oscillator.”</span> <em>Current Biology</em> 13 (16): 1398–408.
</div>
<div id="ref-van2025scaling" class="csl-entry" role="listitem">
Oostrom, Marek J van, Yuting I Li, Wilke HM Meijer, Tomas EJC Noordzij, Charis Fountas, Erika Timmers, Jeroen Korving, Wouter M Thomas, Benjamin D Simons, and Katharina F Sonnen. 2025. <span>“Scaling of Mouse Somitogenesis by Coupling of Cell Cycle to Segmentation Clock Oscillations.”</span> <em>bioRxiv</em>, 2025–01.
</div>
<div id="ref-saga2012mechanism" class="csl-entry" role="listitem">
Saga, Yumiko. 2012. <span>“The Mechanism of Somite Formation in Mice.”</span> <em>Current Opinion in Genetics &amp; Development</em> 22 (4): 331–38.
</div>
<div id="ref-sonnen2018modulation" class="csl-entry" role="listitem">
Sonnen, Katharina F, Volker M Lauschke, Julia Uraji, Henning J Falk, Yvonne Petersen, Maja C Funk, Mathias Beaupeux, Paul François, Christoph A Merten, and Alexander Aulehla. 2018. <span>“Modulation of Phase Shift Between Wnt and Notch Signaling Oscillations Controls Mesoderm Segmentation.”</span> <em>Cell</em> 172 (5): 1079–90.
</div>
<div id="ref-soroldoni2014doppler" class="csl-entry" role="listitem">
Soroldoni, Daniele, David J Jörg, Luis G Morelli, David L Richmond, Johannes Schindelin, Frank Jülicher, and Andrew C Oates. 2014. <span>“A Doppler Effect in Embryonic Pattern Formation.”</span> <em>Science</em> 345 (6193): 222–25.
</div>
<div id="ref-stickney2000somite" class="csl-entry" role="listitem">
Stickney, Heather L, Michael JF Barresi, and Stephen H Devoto. 2000. <span>“Somite Development in Zebrafish.”</span> <em>Developmental Dynamics: An Official Publication of the American Association of Anatomists</em> 219 (3): 287–303.
</div>
<div id="ref-tam1981control" class="csl-entry" role="listitem">
Tam, PPL. 1981. <span>“The Control of Somitogenesis in Mouse Embryos.”</span> <em>Development</em> 65 (Supplement): 103–28.
</div>
<div id="ref-tomka2018travelling" class="csl-entry" role="listitem">
Tomka, Tomas, Dagmar Iber, and Marcelo Boareto. 2018. <span>“Travelling Waves in Somitogenesis: Collective Cellular Properties Emerge from Time-Delayed Juxtacrine Oscillation Coupling.”</span> <em>Progress in Biophysics and Molecular Biology</em> 137: 76–87.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./pattern_practical_2.html" class="pagination-link" aria-label="Turing digit patterns">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Turing digit patterns</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./morpho_intro_text.html" class="pagination-link" aria-label="What is morphogenesis?">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">What is morphogenesis?</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Theoretical Biology, Utrecht University.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was made with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>